lab_4/main.py

```diff
@@ -2,73 +2,306 @@
 Lab 4
 """
 
-from ngrams.ngram_trie import NGramTrie
+## added test for empty inputs, incorrect num inputs
+## added test for save and load model
+
+
+import ast
+from pprint import pprint
+import re
+
+from lab_4.ngrams.ngram_trie import NGramTrie
+from lab_4 import validation
 
 
 def tokenize_by_sentence(text: str) -> tuple:
-    pass
+    validation.ensure_type((text, str))
+    validation.ensure_not_empty((text, ))
```

Comment:
interesting, but i like decorators more)

Reply:
me too, but it became too custom for each function, so i decided to do it more flexible